### PR TITLE
[SPARK-45888][SS] Apply error class framework to State (Metadata) Data Source

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to apply error class framework to the new data source, State (Metadata) Data Source.

### Why are the changes needed?

Error class framework is a standard to represent all exceptions in Spark.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Modified UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @MaxGekk Would you mind taking a look? Thanks in advance!

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3037]: Error classes are coming from specific data source - I don't know how each error class maps to the SQLSTATE, hence I marked all error classes to KD000, SQLSTATE specific to data source.

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:3037]: cc @srielau

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3035]: From Serge: explain how this can happen and how they can mitigate.

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3037]: nit: consider KD006

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3043]: nit: consider 42613

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3047]: instruction to fix the situation from user side.

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3049]: consider 42K03

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3055]: consider XXKST

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3061]: consider 42616

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3067]: [KD](nit: consider KD006)

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3073]: nit: consider KD006

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3079]: nit: consider KD006

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3071]: add explanation, and action item for users

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3077]: add explanation, and action item for users

File: common/utils/src/main/resources/error/error-classes.json
Comment (HeartSaVioR) [inline common/utils/src/main/resources/error/error-classes.json:3085]: consider 42601

**Comment by HeartSaVioR** [comment]
The test failure happened in `Run documentation build` and it doesn't seem to be related.

https://github.com/HeartSaVioR/spark/actions/runs/7005020036/job/19053999542

**Comment by HeartSaVioR** [comment]
@MaxGekk Would you mind taking a look? @srielau helped me to figure out good SQL state values for these error classes, and also made some recommendation. Please let me know if you feel more comfort to let me get sign-off from @srielau before asking you for sign-off. Thanks!

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan @HyukjinKwon to open more chance to be reviewed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:161]: Instead of passing the strings, could you add a sub-class?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala:176]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateScanBuilder.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateScanBuilder.scala:70]: Can't you reuse `SparkException.internalError()`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:111]: Could you use `checkError`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:124]: ditto: Is it possible to use `checkError` to make the tests independent from error formats.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateScanBuilder.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateScanBuilder.scala:70]: How we deal with internal error for data source? Do we think it won't matter if we use a single error class for all of internal errors, among engine and data sources?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala:215]: Please `checkError` too.

**Comment by HeartSaVioR** [comment]
@MaxGekk @beliefer Would you mind taking another round of review? Thanks!

**Review by beliefer — APPROVED** [review]
LGTM. Let @MaxGekk make the final decision.

**Comment by HeartSaVioR** [comment]
Just realized that the failure is relevant. Will update.

**Comment by HeartSaVioR** [comment]
@MaxGekk Mind having a final review and sign-off? Thanks in advance!

**Comment by MaxGekk** [comment]
@HeartSaVioR I will look at it today. Sorry, I haven't had a chance to review it before since I was in a business trip.

**Comment by HeartSaVioR** [comment]
No worries, thanks for the explanation :)

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@HeartSaVioR Could you re-run only the GA [Run / Linters, licenses, dependencies and documentation generation](https://github.com/HeartSaVioR/spark/actions/runs/7096937708/job/19316282574#logs) to make sure everything is ok.

**Comment by HeartSaVioR** [comment]
https://github.com/HeartSaVioR/spark/actions/runs/7111211710/job/19361709611

Kafka test seems to be flaky; it isn't related to the change.

https://github.com/HeartSaVioR/spark/actions/runs/7111211710/job/19358939486

```
Warning, treated as error:
don't know which module to import for autodocumenting 'apply_batch' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)
make: *** [Makefile:35: html] Error 2
                    ------------------------------------------------
      Jekyll 4.3.2   Please append `--trace` to the `build` command 
                     for any additional information or backtrace. 
                    ------------------------------------------------
```

Doesn't seem to be related.

**Comment by HeartSaVioR** [comment]
I'm going to merge as I don't see any relationship with the failure.

**Pull request closed without merging**
